timelock-impl/src/main/java/com/palantir/atlasdb/backup/AtlasBackupResource.java

```diff
@@ -0,0 +1,183 @@
+import com.palantir.atlasdb.timelock.api.CompleteBackupResponse;
+import com.palantir.atlasdb.timelock.api.CompletedBackup;
+import com.palantir.atlasdb.timelock.api.InProgressBackupToken;
+import com.palantir.atlasdb.timelock.api.Namespace;
+import com.palantir.atlasdb.timelock.api.PrepareBackupRequest;
+import com.palantir.atlasdb.timelock.api.PrepareBackupResponse;
+import com.palantir.conjure.java.undertow.lib.UndertowService;
+import com.palantir.lock.v2.IdentifiedTimeLockRequest;
+import com.palantir.lock.v2.LockImmutableTimestampResponse;
+import com.palantir.lock.v2.LockToken;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.logger.SafeLogger;
+import com.palantir.logsafe.logger.SafeLoggerFactory;
+import com.palantir.tokens.auth.AuthHeader;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+public class AtlasBackupResource implements UndertowAtlasBackupClient {
+    private static final SafeLogger log = SafeLoggerFactory.get(AtlasBackupResource.class);
+    private final Function<String, AsyncTimelockService> timelockServices;
+    private final ConjureResourceExceptionHandler exceptionHandler;
+
+    @VisibleForTesting
+    AtlasBackupResource(
+            RedirectRetryTargeter redirectRetryTargeter, Function<String, AsyncTimelockService> timelockServices) {
+        this.exceptionHandler = new ConjureResourceExceptionHandler(redirectRetryTargeter);
+        this.timelockServices = timelockServices;
+    }
+
+    public static UndertowService undertow(
+            RedirectRetryTargeter redirectRetryTargeter, Function<String, AsyncTimelockService> timelockServices) {
+        return AtlasBackupClientEndpoints.of(new AtlasBackupResource(redirectRetryTargeter, timelockServices));
+    }
+
+    public static AtlasBackupClient jersey(
+            RedirectRetryTargeter redirectRetryTargeter, Function<String, AsyncTimelockService> timelockServices) {
+        return new JerseyAtlasBackupClientAdapter(new AtlasBackupResource(redirectRetryTargeter, timelockServices));
+    }
+
+    @Override
+    public ListenableFuture<PrepareBackupResponse> prepareBackup(AuthHeader authHeader, PrepareBackupRequest request) {
+        return handleExceptions(() -> Futures.immediateFuture(prepareBackupInternal(request)));
+    }
+
+    private PrepareBackupResponse prepareBackupInternal(PrepareBackupRequest request) {
+        Set<InProgressBackupToken> preparedBackups = request.getNamespaces().stream()
+                .map(this::prepareBackup)
+                .flatMap(Optional::stream)
+                .collect(Collectors.toSet());
+        return PrepareBackupResponse.of(preparedBackups);
+    }
+
+    Optional<InProgressBackupToken> prepareBackup(Namespace namespace) {
+        try {
+            return Optional.of(tryPrepareBackup(namespace));
+        } catch (Exception ex) {
```

Comment:
Does the above ever throw?

Reply:
looks like not - it must have done in an earlier version of the code.

Reply:
having said that, don't you get some sort of Exception if we fail to reach timelock? Don't we want to handle that here?

Reply:
This is server-side